File: android/app/src/main/java/com/yourcompany/dormmarketplace/YandexAdsModule.kt
package com.yourcompany.dormmarketplace

import android.view.ViewGroup
import android.widget.FrameLayout
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.uimanager.UIManagerModule
import com.yandex.mobile.ads.banner.BannerAdEventListener
import com.yandex.mobile.ads.banner.BannerAdSize
import com.yandex.mobile.ads.banner.BannerAdView
import com.yandex.mobile.ads.common.*
import com.yandex.mobile.ads.interstitial.InterstitialAd
import com.yandex.mobile.ads.interstitial.InterstitialAdEventListener
import com.yandex.mobile.ads.rewarded.Reward
import com.yandex.mobile.ads.rewarded.RewardedAd
import com.yandex.mobile.ads.rewarded.RewardedAdEventListener

class YandexAdsModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private var bannerView: BannerAdView? = null
    private var interstitialAd: InterstitialAd? = null
    private var rewardedAd: RewardedAd? = null
    private var isInterstitialLoaded = false
    private var isRewardedLoaded = false

    init {
        reactContext.addLifecycleEventListener(this)
    }

    override fun getName(): String {
        return "YandexAdsModule"
    }

    private fun sendEvent(eventName: String, params: WritableMap?) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    @ReactMethod
    fun initialize(promise: Promise) {
        try {
            MobileAds.initialize(reactContext) {
                promise.resolve(true)
            }
        } catch (e: Exception) {
            promise.reject("INIT_ERROR", "Failed to initialize Yandex Mobile Ads: ${e.message}", e)
        }
    }

    @ReactMethod
    fun showBanner(adUnitId: String, options: ReadableMap, promise: Promise) {
        val currentActivity = currentActivity ?: run {
            promise.reject("NO_ACTIVITY", "Activity is null")
            return
        }

        try {
            // Clean up existing banner if any
            bannerView?.removeAllViews()
            bannerView?.destroy()
            bannerView = null

            // Get the view tag if provided
            val viewTag = options.getDouble("viewTag").toInt()

            currentActivity.runOnUiThread {
                try {
                    // Create a new banner view
                    bannerView = BannerAdView(reactContext)
                    bannerView?.setAdUnitId(adUnitId)
                    bannerView?.setAdSize(BannerAdSize.fixedSize(reactContext, 320, 50))

                    // Set up event listeners
                    bannerView?.setBannerAdEventListener(object : BannerAdEventListener {
                        override fun onAdLoaded() {
                            val params = Arguments.createMap()
                            params.putString("type", "banner")
                            sendEvent("onAdLoaded", params)
                        }

                        override fun onAdFailedToLoad(error: AdRequestError) {
                            val params = Arguments.createMap()
                            params.putString("type", "banner")
                            params.putString("error", error.description)
                            sendEvent("onAdFailedToLoad", params)
                        }

                        override fun onAdClicked() {
                            val params = Arguments.createMap()
                            params.putString("type", "banner")
                            sendEvent("onAdClicked", params)
                        }

                        override fun onLeftApplication() {
                            // Not sending an event for this
                        }

                        override fun onReturnedToApplication() {
                            // Not sending an event for this
                        }

                        override fun onImpression(impressionData: ImpressionData?) {
                            val params = Arguments.createMap()
                            params.putString("type", "banner")
                            sendEvent("onAdImpression", params)
                        }
                    })

                    // Find the React Native view by tag
                    val uiManager = reactContext.getNativeModule(UIManagerModule::class.java)
                    uiManager?.addUIBlock { nativeViewHierarchyManager ->
                        try {
                            val containerView = nativeViewHierarchyManager.resolveView(viewTag) as? ViewGroup
                            if (containerView != null) {
                                // Add the banner to the container view
                                containerView.removeAllViews()
                                val layoutParams = FrameLayout.LayoutParams(
                                    ViewGroup.LayoutParams.MATCH_PARENT,
                                    ViewGroup.LayoutParams.MATCH_PARENT
                                )
                                bannerView?.layoutParams = layoutParams
                                containerView.addView(bannerView)

                                // Load the ad
                                bannerView?.loadAd(AdRequest.Builder().build())
                                promise.resolve(true)
                            } else {
                                promise.reject("VIEW_NOT_FOUND", "Could not find view with tag $viewTag")
                            }
                        } catch (e: Exception) {
                            promise.reject("VIEW_ERROR", "Error adding banner to view: ${e.message}", e)
                        }
                    }
                } catch (e: Exception) {
                    promise.reject("BANNER_ERROR", "Error creating banner: ${e.message}", e)
                }
            }
        } catch (e: Exception) {
            promise.reject("BANNER_ERROR", "Error showing banner: ${e.message}", e)
        }
    }

    @ReactMethod
    fun hideBanner(promise: Promise) {
        currentActivity?.runOnUiThread {
            try {
                bannerView?.removeAllViews()
                bannerView?.destroy()
                bannerView = null
                promise.resolve(true)
            } catch (e: Exception) {
                promise.reject("BANNER_ERROR", "Error hiding banner: ${e.message}", e)
            }
        }
    }

    @ReactMethod
    fun loadInterstitial(adUnitId: String, promise: Promise) {
        val currentActivity = currentActivity ?: run {
            promise.reject("NO_ACTIVITY", "Activity is null")
            return
        }

        currentActivity.runOnUiThread {
            try {
                // Clean up existing interstitial if any
                interstitialAd?.destroy()
                interstitialAd = null
                isInterstitialLoaded = false

                // Create a new interstitial ad
                interstitialAd = InterstitialAd(reactContext)
                interstitialAd?.setAdUnitId(adUnitId)

                // Set up event listeners
                interstitialAd?.setInterstitialAdEventListener(object : InterstitialAdEventListener {
                    override fun onAdLoaded() {
                        isInterstitialLoaded = true
                        val params = Arguments.createMap()
                        params.putString("type", "interstitial")
                        sendEvent("onAdLoaded", params)
                    }

                    override fun onAdFailedToLoad(error: AdRequestError) {
                        isInterstitialLoaded = false
                        val params = Arguments.createMap()
                        params.putString("type", "interstitial")
                        params.putString("error", error.description)
                        sendEvent("onAdFailedToLoad", params)
                    }

                    override fun onAdShown() {
                        val params = Arguments.createMap()
                        params.putString("type", "interstitial")
                        sendEvent("onAdImpression", params)
                    }

                    override fun onAdDismissed() {
                        isInterstitialLoaded = false
                        val params = Arguments.createMap()
                        params.putString("type", "interstitial")
                        sendEvent("onAdDismissed", params)
                    }

                    override fun onAdClicked() {
                        val params = Arguments.createMap()
                        params.putString("type", "interstitial")
                        sendEvent("onAdClicked", params)
                    }

                    override fun onLeftApplication() {
                        // Not sending an event for this
                    }

                    override fun onReturnedToApplication() {
                        // Not sending an event for this
                    }

                    override fun onImpression(impressionData: ImpressionData?) {
                        // Already handled in onAdShown
                    }
                })

                // Load the ad
                interstitialAd?.loadAd(AdRequest.Builder().build())
                promise.resolve(true)
            } catch (e: Exception) {
                promise.reject("INTERSTITIAL_ERROR", "Error loading interstitial: ${e.message}", e)
            }
        }
    }

    @ReactMethod
    fun showInterstitial(promise: Promise) {
        val currentActivity = currentActivity ?: run {
            promise.reject("NO_ACTIVITY", "Activity is null")
            return
        }

        currentActivity.runOnUiThread {
            try {
                if (interstitialAd != null && isInterstitialLoaded) {
                    interstitialAd?.show()
                    promise.resolve(true)
                } else {
                    promise.reject("NOT_LOADED", "Interstitial ad not loaded")
                }
            } catch (e: Exception) {
                promise.reject("INTERSTITIAL_ERROR", "Error showing interstitial: ${e.message}", e)
            }
        }
    }

    @ReactMethod
    fun isInterstitialLoaded(promise: Promise) {
        promise.resolve(isInterstitialLoaded)
    }

    @ReactMethod
    fun loadRewarded(adUnitId: String, promise: Promise) {
        val currentActivity = currentActivity ?: run {
            promise.reject("NO_ACTIVITY", "Activity is null")
            return
        }

        currentActivity.runOnUiThread {
            try {
                // Clean up existing rewarded if any
                rewardedAd?.destroy()
                rewardedAd = null
                isRewardedLoaded = false

                // Create a new rewarded ad
                rewardedAd = RewardedAd(reactContext)
                rewardedAd?.setAdUnitId(adUnitId)

                // Set up event listeners
                rewardedAd?.setRewardedAdEventListener(object : RewardedAdEventListener {
                    override fun onAdLoaded() {
                        isRewardedLoaded = true
                        val params = Arguments.createMap()
                        params.putString("type", "rewarded")
                        sendEvent("onAdLoaded", params)
                    }

                    override fun onAdFailedToLoad(error: AdRequestError) {
                        isRewardedLoaded = false
                        val params = Arguments.createMap()
                        params.putString("type", "rewarded")
                        params.putString("error", error.description)
                        sendEvent("onAdFailedToLoad", params)
                    }

                    override fun onAdShown() {
                        val params = Arguments.createMap()
                        params.putString("type", "rewarded")
                        sendEvent("onAdImpression", params)
                    }

                    override fun onAdDismissed() {
                        isRewardedLoaded = false
                        val params = Arguments.createMap()
                        params.putString("type", "rewarded")
                        sendEvent("onAdDismissed", params)
                    }

                    override fun onAdClicked() {
                        val params = Arguments.createMap()
                        params.putString("type", "rewarded")
                        sendEvent("onAdClicked", params)
                    }

                    override fun onLeftApplication() {
                        // Not sending an event for this
                    }

                    override fun onReturnedToApplication() {
                        // Not sending an event for this
                    }

                    override fun onImpression(impressionData: ImpressionData?) {
                        // Already handled in onAdShown
                    }

                    override fun onRewarded(reward: Reward) {
                        val params = Arguments.createMap()
                        params.putString("type", "rewarded")
                        params.putInt("amount", reward.amount)
                        params.putString("type", reward.type)
                        sendEvent("onAdRewarded", params)
                    }
                })

                // Load the ad
                rewardedAd?.loadAd(AdRequest.Builder().build())
                promise.resolve(true)
            } catch (e: Exception) {
                promise.reject("REWARDED_ERROR", "Error loading rewarded: ${e.message}", e)
            }
        }
    }

    @ReactMethod
    fun showRewarded(promise: Promise) {
        val currentActivity = currentActivity ?: run {
            promise.reject("NO_ACTIVITY", "Activity is null")
            return
        }

        currentActivity.runOnUiThread {
            try {
                if (rewardedAd != null && isRewardedLoaded) {
                    rewardedAd?.show()
                    promise.resolve(true)
                } else {
                    promise.reject("NOT_LOADED", "Rewarded ad not loaded")
                }
            } catch (e: Exception) {
                promise.reject("REWARDED_ERROR", "Error showing rewarded: ${e.message}", e)
            }
        }
    }

    @ReactMethod
    fun isRewardedLoaded(promise: Promise) {
        promise.resolve(isRewardedLoaded)
    }

    override fun onHostResume() {
        // Nothing to do
    }

    override fun onHostPause() {
        // Nothing to do
    }

    override fun onHostDestroy() {
        bannerView?.destroy()
        bannerView = null
        interstitialAd?.destroy()
        interstitialAd = null
        rewardedAd?.destroy()
        rewardedAd = null
    }
}

File: ios/YandexAdsModule.swift
import Foundation
import YandexMobileAds
import React

@objc(YandexAdsModule)
class YandexAdsModule: RCTEventEmitter {
  
  private var hasListeners = false
  private var bannerView: YMAAdView?
  private var interstitialAd: YMAInterstitialAd?
  private var rewardedAd: YMARewardedAd?
  private var isInterstitialLoaded = false
  private var isRewardedLoaded = false
  
  override init() {
    super.init()
  }
  
  override func supportedEvents() -> [String]! {
    return ["onAdLoaded", "onAdFailedToLoad", "onAdClicked", "onAdImpression", "onAdDismissed", "onAdRewarded"]
  }
  
  override func startObserving() {
    hasListeners = true
  }
  
  override func stopObserving() {
    hasListeners = false
  }
  
  @objc(initialize:reject:)
  func initialize(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      YMAMobileAds.shared().setUserConsent(true)
      YMAMobileAds.shared().activate { error in
        if let error = error {
          reject("INIT_ERROR", "Failed to initialize Yandex Mobile Ads: \(error.localizedDescription)", error)
        } else {
          resolve(true)
        }
      }
    }
  }
  
  @objc(showBanner:options:resolve:reject:)
  func showBanner(adUnitId: String, options: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      // Clean up existing banner if any
      self.bannerView?.removeFromSuperview()
      
      // Get the view tag if provided
      let viewTag = options["viewTag"] as? NSNumber
      
      // Create banner size
      let bannerSize = YMAAdSize.flexibleSize(with: CGSize(width: 320, height: 50))
      
      // Create banner view
      self.bannerView = YMAAdView(adUnitID: adUnitId, adSize: bannerSize)
      self.bannerView?.delegate = self
      
      if let viewTag = viewTag {
        // Find the React Native view by tag
        let rootView = UIApplication.shared.keyWindow?.rootViewController?.view
        let bridge = RCTBridge.current()
        let uiManager = bridge?.module(for: RCTUIManager.self) as? RCTUIManager
        
        uiManager?.addUIBlock { (_, viewRegistry) in
          if let containerView = viewRegistry?[viewTag] as? UIView {
            // Add the banner to the container view
            self.bannerView?.frame = containerView.bounds
            self.bannerView?.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            containerView.addSubview(self.bannerView!)
            
            // Load the ad
            self.bannerView?.loadAd()
            resolve(true)
          } else {
            reject("VIEW_NOT_FOUND", "Could not find view with tag \(viewTag)", nil)
          }
        }
      } else {
        // No view tag provided, reject
        reject("INVALID_VIEW_TAG", "View tag is required", nil)
      }
    }
  }
  
  @objc(hideBanner:reject:)
  func hideBanner(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      self.bannerView?.removeFromSuperview()
      self.bannerView = nil
      resolve(true)
    }
  }
  
  @objc(loadInterstitial:resolve:reject:)
  func loadInterstitial(adUnitId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      self.interstitialAd = YMAInterstitialAd(adUnitID: adUnitId)
      self.interstitialAd?.delegate = self
      self.interstitialAd?.load()
      resolve(true)
    }
  }
  
  @objc(showInterstitial:reject:)
  func showInterstitial(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if let interstitialAd = self.interstitialAd, self.isInterstitialLoaded {
        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
          interstitialAd.present(from: rootViewController)
          resolve(true)
        } else {
          reject("NO_ROOT_VIEW", "Could not find root view controller", nil)
        }
      } else {
        reject("NOT_LOADED", "Interstitial ad not loaded", nil)
      }
    }
  }
  
  @objc(isInterstitialLoaded:reject:)
  func isInterstitialLoaded(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    resolve(self.isInterstitialLoaded)
  }
  
  @objc(loadRewarded:resolve:reject:)
  func loadRewarded(adUnitId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      self.rewardedAd = YMARewardedAd(adUnitID: adUnitId)
      self.rewardedAd?.delegate = self
      self.rewardedAd?.load()
      resolve(true)
    }
  }
  
  @objc(showRewarded:reject:)
  func showRewarded(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if let rewardedAd = self.rewardedAd, self.isRewardedLoaded {
        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
          rewardedAd.present(from: rootViewController)
          resolve(true)
        } else {
          reject("NO_ROOT_VIEW", "Could not find root view controller", nil)
        }
      } else {
        reject("NOT_LOADED", "Rewarded ad not loaded", nil)
      }
    }
  }
  
  @objc(isRewardedLoaded:reject:)
  func isRewardedLoaded(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    resolve(self.isRewardedLoaded)
  }
  
  // Required for RCTEventEmitter
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
}

// MARK: - YMAAdViewDelegate
extension YandexAdsModule: YMAAdViewDelegate {
  func adViewDidLoad(_ adView: YMAAdView) {
    if hasListeners {
      sendEvent(withName: "onAdLoaded", body: ["type": "banner"])
    }
  }
  
  func adViewDidFailLoading(_ adView: YMAAdView, error: Error) {
    if hasListeners {
      sendEvent(withName: "onAdFailedToLoad", body: ["type": "banner", "error": error.localizedDescription])
    }
  }
  
  func adViewDidClick(_ adView: YMAAdView) {
    if hasListeners {
      sendEvent(withName: "onAdClicked", body: ["type": "banner"])
    }
  }
  
  func adViewWillLeaveApplication(_ adView: YMAAdView) {
    // Not sending an event for this
  }
  
  func adViewDidTrackImpression(_ adView: YMAAdView) {
    if hasListeners {
      sendEvent(withName: "onAdImpression", body: ["type": "banner"])
    }
  }
}

// MARK: - YMAInterstitialAdDelegate
extension YandexAdsModule: YMAInterstitialAdDelegate {
  func interstitialAdDidLoad(_ interstitialAd: YMAInterstitialAd) {
    self.isInterstitialLoaded = true
    if hasListeners {
      sendEvent(withName: "onAdLoaded", body: ["type": "interstitial"])
    }
  }
  
  func interstitialAdDidFailToLoad(_ interstitialAd: YMAInterstitialAd, error: Error) {
    self.isInterstitialLoaded = false
    if hasListeners {
      sendEvent(withName: "onAdFailedToLoad", body: ["type": "interstitial", "error": error.localizedDescription])
    }
  }
  
  func interstitialAdDidClick(_ interstitialAd: YMAInterstitialAd) {
    if hasListeners {
      sendEvent(withName: "onAdClicked", body: ["type": "interstitial"])
    }
  }
  
  func interstitialAdDidShow(_ interstitialAd: YMAInterstitialAd) {
    if hasListeners {
      sendEvent(withName: "onAdImpression", body: ["type": "interstitial"])
    }
  }
  
  func interstitialAdDidDismiss(_ interstitialAd: YMAInterstitialAd) {
    self.isInterstitialLoaded = false
    if hasListeners {
      sendEvent(withName: "onAdDismissed", body: ["type": "interstitial"])
    }
  }
}

// MARK: - YMARewardedAdDelegate
extension YandexAdsModule: YMARewardedAdDelegate {
  func rewardedAdDidLoad(_ rewardedAd: YMARewardedAd) {
    self.isRewardedLoaded = true
    if hasListeners {
      sendEvent(withName: "onAdLoaded", body: ["type": "rewarded"])
    }
  }
  
  func rewardedAdDidFailToLoad(_ rewardedAd: YMARewardedAd, error: Error) {
    self.isRewardedLoaded = false
    if hasListeners {
      sendEvent(withName: "onAdFailedToLoad", body: ["type": "rewarded", "error": error.localizedDescription])
    }
  }
  
  func rewardedAdDidClick(_ rewardedAd: YMARewardedAd) {
    if hasListeners {
      sendEvent(withName: "onAdClicked", body: ["type": "rewarded"])
    }
  }
  
  func rewardedAdDidShow(_ rewardedAd: YMARewardedAd) {
    if hasListeners {
      sendEvent(withName: "onAdImpression", body: ["type": "rewarded"])
    }
  }
  
  func rewardedAdDidDismiss(_ rewardedAd: YMARewardedAd) {
    self.isRewardedLoaded = false
    if hasListeners {
      sendEvent(withName: "onAdDismissed", body: ["type": "rewarded"])
    }
  }
  
  func rewardedAd(_ rewardedAd: YMARewardedAd, didReward reward: YMAReward) {
    if hasListeners {
      sendEvent(withName: "onAdRewarded", body: [
        "type": "rewarded",
        "amount": reward.amount,
        "type": reward.type
      ])
    }
  }
}